migrate_cpp/cpp_refactoring/rewriter.h

```diff
@@ -0,0 +1,205 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_MIGRATE_CPP_CPP_REFACTORING_REWRITER_H_
+#define CARBON_MIGRATE_CPP_CPP_REFACTORING_REWRITER_H_
+
+#include <optional>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/ASTTypeTraits.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendAction.h"
+#include "llvm/ADT/DenseMap.h"
+#include "migrate_cpp/cpp_refactoring/output_segment.h"
+
+namespace Carbon {
+
+// `OutputWriter` is ressponsible for traversing the tree of `OutputSegment`s
+// and writing the correct data to its member `output`.
+class OutputWriter {
+  struct Empty {
+    friend bool operator==(Empty, Empty) { return true; }
+  };
+  struct Tombstone {
+    friend bool operator==(Tombstone, Tombstone) { return true; }
+  };
+  using KeyType =
+      std::variant<clang::DynTypedNode, clang::TypeLoc, Empty, Tombstone>;
+
+  struct KeyInfo {
```

Comment:
Can you add comments about this? I assume it's required for DenseMap's API, but a brief comment to that effect would help me since I'm not too familiar with llvm's APIs.

Reply:
Done.